SSR can't resolve builtin modules that start with underscore

### Describe the bug

Some packages (like `@prisma/client`) require the builtin `_http_common` module (or others that start with an underscore) which causes `ssrLoadModule` to throw `Error: Cannot find module '_http_common' imported from ...` error. It seems to be caused by the `builtin-modules` package that Vite uses for this purpose filtering out modules that start with an underscore (changing the relevant regex in the transpiled code fixes the problem).

### Reproduction

- Clone the repo and install the dependencies: https://github.com/cyco130/vite-ssr-http-common-issue
- `node ssr-module.mjs` works but `node .` throws trying to load the same module with `ssrLoadModule`.

### System Info

```shell
System:
    OS: Linux 5.4 Linux Mint 20.2 (Uma)
    CPU: (8) x64 Intel(R) Core(TM) i7-3630QM CPU @ 2.40GHz
    Memory: 1.70 GB / 15.53 GB
    Container: Yes
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 14.18.1 - /usr/local/bin/node
    Yarn: 1.22.11 - ~/.npm-global/bin/yarn
    npm: 7.24.1 - ~/.npm-global/bin/npm
  Browsers:
    Chrome: 96.0.4664.45
    Firefox: 94.0
```


### Used Package Manager

npm

### Logs

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/vue-next instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite/package.json
+++ b/packages/vite/package.json
@@ -79,7 +79,6 @@
     "acorn": "^8.6.0",
     "acorn-class-fields": "^1.0.0",
     "acorn-static-class-features": "^1.0.0",
-    "builtin-modules": "^3.2.0",
     "cac": "6.7.9",
     "chalk": "^4.1.2",
     "chokidar": "^3.5.2",

--- a/packages/vite/src/node/utils.ts
+++ b/packages/vite/src/node/utils.ts
@@ -38,7 +38,7 @@ export const normalizeId = (id: string): string =>
   id.replace(/(\s*>\s*)/g, ' > ')
   
 const builtins = [...builtinModules]
-if (!builtins.includes('fs/promises')) builtins.push('fs/promises')
+if (typeof jest === 'undefined' && !builtins.includes('fs/promises')) builtins.push('fs/promises')
 builtins.forEach(builtin => { builtins.push(`node:${builtin}`) })
 
 export function isBuiltin(id: string): boolean {

--- a/packages/vite/src/node/utils.ts
+++ b/packages/vite/src/node/utils.ts
@@ -56,7 +56,7 @@ const builtins = new Set([
 ])
 
 export function isBuiltin(id: string): boolean {
-  return builtins.has(id.replace('node:', ''))
+  return builtins.has(id.replace(/^node:/, ''))
 }
 
 export function moduleListContains(

--- a/packages/vite/src/node/utils.ts
+++ b/packages/vite/src/node/utils.ts
@@ -38,11 +38,25 @@ export const normalizeId = (id: string): string =>
   id.replace(/(\s*>\s*)/g, ' > ')
 
 //TODO: revisit later to see if the edge case that "compiling using node v12 code to be run in node v16 in the server" is what we intend to support.
-const builtins = [...builtinModules, 'assert/strict','diagnostics_channel','dns/promises','fs/promises','path/posix','path/win32' ,'readline/promises','stream/consumers','stream/promises','stream/web','timers/promises','util/types','wasi']
-builtins.forEach(builtin => { builtins.push(`node:${builtin}`) })
+const builtins = new Set([
+  ...builtinModules,
+  'assert/strict',
+  'diagnostics_channel',
+  'dns/promises',
+  'fs/promises',
+  'path/posix',
+  'path/win32',
+  'readline/promises',
+  'stream/consumers',
+  'stream/promises',
+  'stream/web',
+  'timers/promises',
+  'util/types',
+  'wasi'
+])
 
 export function isBuiltin(id: string): boolean {
-  return builtins.includes(id)
+  return builtins.has(id.replace('node:', ''))
 }
 
 export function moduleListContains(

--- a/packages/vite/src/node/utils.ts
+++ b/packages/vite/src/node/utils.ts
@@ -12,7 +12,7 @@ import {
   ENV_PUBLIC_PATH
 } from './constants'
 import resolve from 'resolve'
-import builtins from 'builtin-modules'
+import { builtinModules } from 'module'
 import { FSWatcher } from 'chokidar'
 import remapping from '@ampproject/remapping'
 import {
@@ -38,8 +38,9 @@ export const normalizeId = (id: string): string =>
   id.replace(/(\s*>\s*)/g, ' > ')
 
 export function isBuiltin(id: string): boolean {
-  const deepMatch = id.match(deepImportRE)
-  id = deepMatch ? deepMatch[1] || deepMatch[2] : id
+  const builtins = [...builtinModules]
+  if (!builtins.includes('fs/promises')) builtins.push('fs/promises')
+  builtins.forEach(builtin => { builtins.push(`node:${builtin}`) })
   return builtins.includes(id)
 }
 

--- a/packages/vite/src/node/utils.ts
+++ b/packages/vite/src/node/utils.ts
@@ -40,6 +40,7 @@ export const normalizeId = (id: string): string =>
 export function isBuiltin(id: string): boolean {
   const builtins = [...builtinModules]
   if (!builtins.includes('fs/promises')) builtins.push('fs/promises')
+  if (!builtins.includes('diagnostics_channel')) builtins.push('diagnostics_channel') 
   builtins.forEach(builtin => { builtins.push(`node:${builtin}`) })
   return builtins.includes(id)
 }

--- a/packages/vite/src/node/utils.ts
+++ b/packages/vite/src/node/utils.ts
@@ -36,9 +36,9 @@ export const flattenId = (id: string): string =>
 
 export const normalizeId = (id: string): string =>
   id.replace(/(\s*>\s*)/g, ' > ')
-  
-const builtins = [...builtinModules]
-if (typeof jest !== 'undefined' && !builtins.includes('fs/promises')) builtins.push('fs/promises')
+
+//TODO: revisit later to see if the edge case that "compiling using node v12 code to be run in node v16 in the server" is what we intend to support.
+const builtins = [...builtinModules, 'assert/strict','diagnostics_channel','dns/promises','fs/promises','path/posix','path/win32' ,'readline/promises','stream/consumers','stream/promises','stream/web','timers/promises','util/types','wasi']
 builtins.forEach(builtin => { builtins.push(`node:${builtin}`) })
 
 export function isBuiltin(id: string): boolean {

--- a/packages/vite/src/node/utils.ts
+++ b/packages/vite/src/node/utils.ts
@@ -36,12 +36,11 @@ export const flattenId = (id: string): string =>
 
 export const normalizeId = (id: string): string =>
   id.replace(/(\s*>\s*)/g, ' > ')
+  
+const builtins = [...builtinModules]
+builtins.forEach(builtin => { builtins.push(`node:${builtin}`) })
 
 export function isBuiltin(id: string): boolean {
-  const builtins = [...builtinModules]
-  if (!builtins.includes('fs/promises')) builtins.push('fs/promises')
-  if (!builtins.includes('diagnostics_channel')) builtins.push('diagnostics_channel') 
-  builtins.forEach(builtin => { builtins.push(`node:${builtin}`) })
   return builtins.includes(id)
 }
 

--- a/packages/vite/src/node/utils.ts
+++ b/packages/vite/src/node/utils.ts
@@ -38,6 +38,7 @@ export const normalizeId = (id: string): string =>
   id.replace(/(\s*>\s*)/g, ' > ')
   
 const builtins = [...builtinModules]
+if (!builtins.includes('fs/promises')) builtins.push('fs/promises')
 builtins.forEach(builtin => { builtins.push(`node:${builtin}`) })
 
 export function isBuiltin(id: string): boolean {

--- a/packages/vite/src/node/utils.ts
+++ b/packages/vite/src/node/utils.ts
@@ -38,7 +38,7 @@ export const normalizeId = (id: string): string =>
   id.replace(/(\s*>\s*)/g, ' > ')
   
 const builtins = [...builtinModules]
-if (typeof jest === 'undefined' && !builtins.includes('fs/promises')) builtins.push('fs/promises')
+if (typeof jest !== 'undefined' && !builtins.includes('fs/promises')) builtins.push('fs/promises')
 builtins.forEach(builtin => { builtins.push(`node:${builtin}`) })
 
 export function isBuiltin(id: string): boolean {

